File: src/renderers.js
// Each renderer takes the same per-bar data (already computed: position,
// height, color) and just differs in how it paints it — swapping display
// mode never touches the audio/color logic upstream.

export function drawBarsNormal(ctx, h, bars, barWidth) {
  for (const bar of bars) {
    ctx.fillStyle = `rgb(${bar.r}, ${bar.g}, ${bar.b})`
    ctx.fillRect(bar.x, h - bar.height, barWidth * 0.8, bar.height)
  }
}

const PIXEL_GROUP = 4 // bars grouped together per chunky "pixel" column
const LEVELS = 10 // discrete height steps, like a retro LED EQ
const COLOR_STEP = 36 // quantized color depth

function quantize(v, step) {
  return Math.min(255, Math.round(v / step) * step)
}

export function drawBars8Bit(ctx, h, bars, barWidth) {
  const groupWidth = barWidth * PIXEL_GROUP
  const segmentHeight = h / LEVELS

  for (let g = 0; g < bars.length; g += PIXEL_GROUP) {
    const group = bars.slice(g, g + PIXEL_GROUP)
    if (group.length === 0) continue

    let sumHeight = 0
    let sumR = 0
    let sumG = 0
    let sumB = 0
    for (const bar of group) {
      sumHeight += bar.height
      sumR += bar.r
      sumG += bar.g
      sumB += bar.b
    }
    const avgHeight = sumHeight / group.length
    const r = quantize(sumR / group.length, COLOR_STEP)
    const gCol = quantize(sumG / group.length, COLOR_STEP)
    const b = quantize(sumB / group.length, COLOR_STEP)
    const levelCount = Math.max(0, Math.round((avgHeight / h) * LEVELS))

    ctx.fillStyle = `rgb(${r}, ${gCol}, ${b})`
    const x = group[0].x
    for (let lvl = 0; lvl < levelCount; lvl++) {
      const y = h - (lvl + 1) * segmentHeight
      ctx.fillRect(x, y + 2, groupWidth * 0.85, segmentHeight - 4)
    }
  }
}

export function drawCurveArea(ctx, w, h, bars) {
  if (bars.length === 0) return

  let sumR = 0
  let sumG = 0
  let sumB = 0
  for (const bar of bars) {
    sumR += bar.r
    sumG += bar.g
    sumB += bar.b
  }
  const n = bars.length
  const avgR = Math.round(sumR / n)
  const avgG = Math.round(sumG / n)
  const avgB = Math.round(sumB / n)

  ctx.beginPath()
  ctx.moveTo(0, h)
  ctx.lineTo(bars[0].x, h - bars[0].height)
  for (let i = 1; i < bars.length; i++) {
    const prev = bars[i - 1]
    const curr = bars[i]
    const midX = (prev.x + curr.x) / 2
    const midY = h - (prev.height + curr.height) / 2
    ctx.quadraticCurveTo(prev.x, h - prev.height, midX, midY)
  }
  const last = bars[bars.length - 1]
  ctx.lineTo(w, h - last.height)
  ctx.lineTo(w, h)
  ctx.closePath()

  const gradient = ctx.createLinearGradient(0, 0, 0, h)
  gradient.addColorStop(0, `rgba(${avgR}, ${avgG}, ${avgB}, 0.85)`)
  gradient.addColorStop(1, `rgba(${avgR}, ${avgG}, ${avgB}, 0.05)`)
  ctx.fillStyle = gradient
  ctx.fill()

  ctx.strokeStyle = `rgb(${avgR}, ${avgG}, ${avgB})`
  ctx.lineWidth = 2
  ctx.stroke()
}

File: macos/Sources/VisualizeAudio/Visualizer/Renderers.swift
import SwiftUI

// Direct port of src/renderers.js. Each renderer takes the same per-bar data
// (already computed: position, height, color) and just differs in how it
// paints it — swapping display mode never touches the audio/color logic
// upstream. Keep in sync with renderers.js.

enum DisplayMode: String, CaseIterable {
    case normal = "Normal"
    case eightBit = "8-Bit"
    case curve = "Curve"
}

private func color(_ bar: Bar) -> Color {
    Color(red: bar.r / 255, green: bar.g / 255, blue: bar.b / 255)
}

func drawBarsNormal(_ context: GraphicsContext, height: Double, bars: [Bar], barWidth: Double) {
    for bar in bars {
        let rect = CGRect(x: bar.x, y: height - bar.height, width: barWidth * 0.8, height: bar.height)
        context.fill(Path(rect), with: .color(color(bar)))
    }
}

private let pixelGroup = 4 // bars grouped together per chunky "pixel" column
private let levels = 10 // discrete height steps, like a retro LED EQ
private let colorStep = 36.0 // quantized color depth

private func quantize(_ v: Double, _ step: Double) -> Double {
    min(255, (v / step).rounded() * step)
}

func drawBars8Bit(_ context: GraphicsContext, height: Double, bars: [Bar], barWidth: Double) {
    let groupWidth = barWidth * Double(pixelGroup)
    let segmentHeight = height / Double(levels)

    var g = 0
    while g < bars.count {
        let group = bars[g..<min(g + pixelGroup, bars.count)]
        if group.isEmpty { g += pixelGroup; continue }

        var sumHeight = 0.0, sumR = 0.0, sumG = 0.0, sumB = 0.0
        for bar in group {
            sumHeight += bar.height
            sumR += bar.r
            sumG += bar.g
            sumB += bar.b
        }
        let count = Double(group.count)
        let avgHeight = sumHeight / count
        let r = quantize(sumR / count, colorStep)
        let gCol = quantize(sumG / count, colorStep)
        let b = quantize(sumB / count, colorStep)
        let levelCount = max(0, Int((avgHeight / height * Double(levels)).rounded()))

        let fillColor = Color(red: r / 255, green: gCol / 255, blue: b / 255)
        let x = group.first!.x
        for lvl in 0..<levelCount {
            let y = height - Double(lvl + 1) * segmentHeight
            let rect = CGRect(x: x, y: y + 2, width: groupWidth * 0.85, height: segmentHeight - 4)
            context.fill(Path(rect), with: .color(fillColor))
        }
        g += pixelGroup
    }
}

func drawCurveArea(_ context: GraphicsContext, width: Double, height: Double, bars: [Bar]) {
    guard !bars.isEmpty else { return }

    var sumR = 0.0, sumG = 0.0, sumB = 0.0
    for bar in bars {
        sumR += bar.r
        sumG += bar.g
        sumB += bar.b
    }
    let n = Double(bars.count)
    let avgR = (sumR / n).rounded()
    let avgG = (sumG / n).rounded()
    let avgB = (sumB / n).rounded()

    var path = Path()
    path.move(to: CGPoint(x: 0, y: height))
    path.addLine(to: CGPoint(x: bars[0].x, y: height - bars[0].height))
    for i in 1..<bars.count {
        let prev = bars[i - 1]
        let curr = bars[i]
        let midX = (prev.x + curr.x) / 2
        let midY = height - (prev.height + curr.height) / 2
        path.addQuadCurve(to: CGPoint(x: midX, y: midY), control: CGPoint(x: prev.x, y: height - prev.height))
    }
    let last = bars[bars.count - 1]
    path.addLine(to: CGPoint(x: width, y: height - last.height))
    path.addLine(to: CGPoint(x: width, y: height))
    path.closeSubpath()

    let gradient = Gradient(stops: [
        .init(color: Color(red: avgR / 255, green: avgG / 255, blue: avgB / 255).opacity(0.85), location: 0),
        .init(color: Color(red: avgR / 255, green: avgG / 255, blue: avgB / 255).opacity(0.05), location: 1),
    ])
    context.fill(
        path,
        with: .linearGradient(gradient, startPoint: CGPoint(x: 0, y: 0), endPoint: CGPoint(x: 0, y: height))
    )
    context.stroke(path, with: .color(Color(red: avgR / 255, green: avgG / 255, blue: avgB / 255)), lineWidth: 2)
}
